### PR TITLE
Use explicit mixed for global array variables

### PR DESCRIPTION
**Pull request by herndlm**
Refs: https://github.com/phpstan/phpstan/discussions/7450

well, let's see how much this will break..

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Definitely needs to be a bleedingEdge tag :) Don't forget to update LazyScopeFactory + DirectScopeFactory.

File: tests/PHPStan/Rules/Functions/CallToFunctionParametersRuleTest.php [hunk 1]
```diff
@@ -1077,4 +1077,19 @@ public function testBug6987(): void
 		$this->analyse([__DIR__ . '/data/bug-6987.php'], []);
 	}
 
+	public function testDiscussion7450(): void
+	{
+		$this->checkExplicitMixed = true;
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/Functions/CallToFunctionParametersRuleTest.php:1088]: Please test it with `false` too.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
Perfect, thank you!

**Comment by gnutix** [comment]
Hey there. Just a question @herndlm :

> I try to explain to colleagues that we should handle this type-safe

How exactly do you do that in this scenario ? Storing the value of the global array you want to use into a variable and using `@var` on the variable, or doing `is_*()` (string, numeric, ...) kind of checks on the value of the global array before using it ?

**Comment by ondrejmirtes** [comment]
Developers should never use `@var`, especially not on an untrusted user input 😊

The right approach now is cuyz/valinor, or azjezz/psl.

**Comment by herndlm** [comment]
> How exactly do you do that in this scenario ? Storing the value of the global array you want to use into a variable and using `@var` on the variable, or doing `is_*()` (string, numeric, ...) kind of checks on the value of the global array before using it ?

yeah, basically just https://phpstan.org/writing-php-code/narrowing-types. We just had some simple primitive type fields/vars from $_POST, just as the playground-snippet from the linked discussion.

for more complex array shapes I would definitely also try to use what Ondrej suggested in the previous comment. Never used those libs, but they are on my list :)

**Pull request merged**